Layer multiselect support (WIP)

saif-ellafi opened the pull request:
## PR Checklist (Required)
Please ensure your PR meets the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/hm21/pro_image_editor/blob/stable/CONTRIBUTING.md#style-guides
- [ ] I have run `dart format .` in the terminal and committed any changes.
- [ ] I have run `dart analyze .` in the terminal and addressed any issues.
- [ ] I have run `flutter test` in the terminal and addressed any issues.
- [ ] I have pulled from the stable branch before committing.
- [ ] I have updated the `CHANGELOG.md` file with my changes (For the version, you can use X.X.X, I will later bump it after it's merged).
- [ ] The PR title follows the conventional commit style (e.g., `feat(paint-editor): add new triangle shape`).
- [ ] If this PR introduces breaking changes, I have verified that there is no way to implement the changes without them.

## PR Checklist (Optional)
- [ ] Tests have been added for the changes.
- [ ] Documentation has been added/updated.

## PR Type
Please indicate the types of changes this PR introduces by checking the relevant boxes:

- [ ] Bugfix
- [X] Feature
- [ ] Performance
- [ ] Refactor (no functional or API changes)
- [ ] Code style updates (e.g., formatting, local variables)
- [ ] Documentation updates
- [ ] CI-related changes
- [ ] Other... Please describe:

## Current Behavior
The goal is to allow selection of multiple layers and being able to move them around.
Also, to be able to persist "groups" so layers of the same groupId could be persisted as a group, and on selection, they get selected together.

Issue Number: [Discussion](https://github.com/hm21/pro_image_editor/discussions/569)

## New Behavior
- SHIFT or CTRL + Click will multi-select layers (working fine?)
- Moving any of the selected layer moves the selected layers (not working as expected - gets stuck in moving state sometimes)
- Backend implementation of a groupId that can be persisted (wip/not working as expected)
- Undo/Redo history moves moved layers together (wip/not working as expected)
- For Mobile/Touch Screen implement long-press to do a "Multi-Select" (not implemented)
- Allow an application-side flag to toggle multi-select mode on/off (could be useful for device instead of long-press)

## Comments
NOT ready, nor finished, I am testing this and helping myself with copilot, debugging and trying to learn how the layer management works. It works more or less fine, but there's a few issues in history management, I could not fix the undoAction() not solving properly on multi-select layers. Click outside to de-select works fine.

I might not be able to do it right, but for the complexity, it looks like the current plan is/should not be a big amount of effort or complexity to implement properly under this approach (multiple active layers, groupId, and move together).

[Screencast_20250725_024912.webm](https://github.com/user-attachments/assets/6ae63b9b-71de-4d4c-845f-58e792a28fa4)

## Breaking Changes?
Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

saif-ellafi committed:
feat(multiselect): first version attempt

saif-ellafi committed:
Merge branch 'stable' into layer-multiselect-support

hm21 committed:
fix(minified-keys): add 'groupId' to minified layer keys

hm21 committed:
feat(keyboard-service): implement KeyboardService for key state management

hm21 committed:
feat(callbacks): add callback for multiple selected layers

hm21 commented:
@saif-ellafi Wow, it already looks great, and I’ve just played around with it a bit. I’ll also commit some small improvements that extend the callbacks and make the groupId exportable/importable.

I quickly created a Kanban board [here](https://github.com/users/hm21/projects/12/views/1) so we can manage the open tasks before the release. For now, I just added the things you mentioned above but feel free to add more if you spot anything else.

I’ll now take a look at the Undo/Redo issue you pointed out.


----

If anyone’s interested in collaborating on this feature, feel free to comment below, I’ll give you write access to the Kanban board.

saif-ellafi commented:
Woah! Thank you so much! catching up with all changes and testing now!

hm21 committed:
feat(layer-selection): replace legacy layer selection system with new implementation

- Updated LayerInteractionManager to handle multiple selected layers, including methods for adding and removing multiple layers.
- Replaced single layer references with lists in methods for scaling and rotating layers.
- Enhanced snapping functionality by tracking rotation and scale factors for each layer individually.
- Modified MainEditorInteractiveContent and MainEditorLayers to utilize the new multi-select capabilities.
- Adjusted UI components to reflect changes in layer selection logic, ensuring proper interaction feedback.
- Improved performance by reducing redundant state updates and optimizing layer interaction checks.

hm21 commented:
@saif-ellafi I updated the layer selection system. There was a lot more to change than I thought because the old system also used a `selectedLayerIndex`, which I also had to completely remove. 
However, I think it should work fine now, and the undo/redo functions should also work properly. Keep in mind that I just finished it on desktop devices. I will implement it on mobile platforms now.

saif-ellafi commented:
Hi @hm21 - will test! I hope we didn't clash on the same part of the code, I've been a few hours trying to make all selected layers to move together, because currently (this morning), they only move in groups if you start the movement from the main first selected layer. Perhaps you fixed that already?

Second, not sure if a side-effect, but I notice that if I zoom in into the canvas after moving a group of layers, one of the layers begin to scale up beyond after I hit the limit of the zoom in.

Third, for mobile, doing Long-Press I think is super clever, I also had the idea reserved for the `enableMultiSelect` mode as a state where any layer click would append the selection (or de-select if selected from the group), but this might not be needed if mobile long-press is good enough.

saif-ellafi commented:
Just tested @hm21 - and indeed you fixed that group movement! However the scaling issue you still have, I think I have the fix, checking it out

hm21 commented:
@saif-ellafi  Good that the group movement works by you also correctly now ;) 

Perfect I will check your fix for the scaling issue after you commit. 

I will move forward with the mobile selection. It will be a bit buggy because of the changes I just made.

saif-ellafi committed:
fix(multi-layer-zoom): absorb zoom in

saif-ellafi commented:
@hm21 - fixed zoom in and out after multi-layer selection (I hope)

hm21 committed:
fix(mobile-selection): show overlay on mobile only when enabled in config

hm21 committed:
fix(main-editor-layers): clear selected layers when layers are not selectable

saif-ellafi commented:
@hm21 by the way, I don't think that group scaling/rotating selected layers should be done (or at least not by default). I think scaling and rotating should always be on the active layer (where the user pulls the handle interaction buttons)

Of course it wouldn't hurt to have but I guess with a Config (default true/false depending on what feels more intuitive?). At least for all the use cases I am thinking, group scaling and rotation is not really useful, in the end the user will want to adjust individually I think.

I could always be wrong of course, and perhaps everything should happen in groups when multi-selected. But I was thinking perhaps if multi-selection drew a single large box for all selected layers, in the current model, individual scaling feels a bit better for me. Again, a setting could let user decide but add complexity.

Will now proceed to check and test persisting groupId and how could application side work with it

hm21 commented:
@saif-ellafi I just tested your fix for layer scaling, and it works fine now :)

> Of course it wouldn't hurt to have but I guess with a Config (default true/false depending on what feels more intuitive?

I agree with you. I will check how complicated it is to extend with the config flag, as you suggested. Only if it's easy, I will implement it.

saif-ellafi commented:
When a layer has `enableMove == false` and a multi-selection attempts to move a group, all layers move except the ones that are "locked". Looks good!

hm21 committed:
feat(layer-selection): implement long press layer selection

hm21 committed:
feat(main-editor): enable multi-select mode without key modifiers

hm21 commented:
Btw it was a good idea from you to add the `enableMultiSelectMode` function. I just implemented it now, and it makes it really easy for people to select from a lot of layers.

hm21 committed:
fix(layer-widget): prevent long press detection on layer changes

hm21 committed:
fix(main-editor): disable multi-select mode by default and adjust layer rotation logic

saif-ellafi commented:
Amazing! I got stuck on my end, I am trying to add an Interaction Button to make use of the groupId

```dart
          LayerInteractionButton(
            rotation: -layer.rotation,
            onTap: () {
              if (layer.groupId != null) {
                print('Ungrouping layer ${layer.id}');
                interactions.ungroup();
              } else {
                print('Grouping layer ${layer.id}');
                interactions.group();
              }
            },
            buttonRadius: _getInteractionRadius(layer),
            cursor: SystemMouseCursors.click,
            icon: layer.groupId == null ? Icons.group : Icons.group_off,
            color: Colors.black,
            background: Colors.white,
          ),
```

But I can't get it to work. The groupId resets back to null as soon as I use the interaction, I think the other layer is being de-selected. Checking...

hm21 commented:
It sounds like there's a problem with the absorbed gesture. I'm not sure if this will help, but wrapping it with a `GestureInterceptor` might fix it?

hm21 committed:
fix(layer-interaction): resolve broken layer scale/rotate function

hm21 commented:
The Scale/Rotate button should now work correctly with multiselection. 
It's a bit tricky to transform all layers through a single interaction button, so for now I’m not implementing that. As you mentioned, it’s nice to have but not essential, and I also don't think anyone will use that option.

saif-ellafi commented:
Got it! About to push in a moment

saif-ellafi committed:
fix(layer-interaction): persist groupId in copy, implement group and ungroup

saif-ellafi commented:
@hm21 Pushed - I managed to have a working grouping interaction function set (it is not added by default to the interactions). 

I included copy code to persist the groupId, the interaction functions (out of the box, non-overrideable), persists the group step in history, select all layers of same group Id when selecting a layer (in main_editor_layers.dart).

Code Example to use it is same as above

saif-ellafi commented:
[Screencast_20250725_161517.webm](https://github.com/user-attachments/assets/042b73d7-4980-4baa-9528-5dd356b5ea8c)

example usage

hm21 committed:
feat(layer-group-example): implement new layer grouping example

hm21 commented:
Awesome, I tested it too and it works great. I’ve now added a new example `layer_grouping_example.dart`, which you can find in the demo under `Layers => Group-Selection`. It also includes the ability to group layers, just like you showed above.

hm21 commented:
Btw it looks like we're done with the implementation, right? If so, I think I’ll go ahead and prepare a prerelease on pub.dev so others can easily test it before we fully release it.

saif-ellafi commented:
I see two "small" things that I am looking into asap, that seem necessary:

1. Grouping works fine if you "single tap" the layers of the group. But if you skip ahead and directly "move" the layer, it won't select&move the entire group! (needs fixing, have not yet looked at it)
2. My app has a Copy Layer interaction (maybe one day it would be better in PIE), which copy-paste a layer a few pixels to the bottom-right. I am testing that a copy-paste of a group, copy-pastes both selected layers, I am testing this and might push something in a few minutes to make a copy of the groupId into a different group, then go to point 1 unless you find the issue first!

hm21 commented:
I’ll quickly write the changelog with the breaking changes and update the README. After that, I’ll take a look at point 1, unless you’ve already found the issue by then.

hm21 committed:
chore: update changelog for version 11.0.0 with multi-select and grouping features

hm21 committed:
feat(layer-interaction): add keyboard and long-press multi-selection options

hm21 committed:
docs(layer-interaction): add missing doc comments

saif-ellafi commented:
Hmm, I am not sure in which commit, but there's a scale issue again :( when I move a layer, then zoom in/zoom out beyond limits, layers are scaling with me. Checking...

[Screencast_20250725_170919.webm](https://github.com/user-attachments/assets/e319ed61-cd1b-4818-9ef6-427fa6889b9b)

hm21 commented:
Ah, I got it. Wait a second, I know where the problem is!

saif-ellafi committed:
fix(multi-layer): import groupId from state

saif-ellafi committed:
Merge branch 'layer-multiselect-support' of https://github.com/saif-ellafi/pro_image_editor into layer-multiselect-support

saif-ellafi commented:
I pushed a bugfix to load from saved states the groupId..

Copy paste issue can be solved from application side, so all good, I think, just checking some math to paste and respect original offsets.. getting close :)

hm21 committed:
fix(interactive-viewer): resolve scale issue when editor zoom is enabled

hm21 committed:
Merge branch 'layer-multiselect-support' of https://github.com/saif-ellafi/pro_image_editor into pr/saif-ellafi/581

hm21 commented:
Okay, I know what the problem is with Issue 1 now. I'll fix it.

saif-ellafi commented:
I was thinking whether it could be left broken as a way for the users to move something individually when grouped :D but all good, I think better fix yes, and let the users "ungroup" first. Nice! I hope no other hidden bugs are there, checking and testing

saif-ellafi commented:
Overall all good, I only have one issue which is when copying layers of different types together, they are not respecting their relative offsets, but well, could be for later...

(application side)
```dart
void copyAllSelectedLayers() {
  final editorState = editorKey.currentState;
  if (editorState == null) {
    return;
  }

  final selectedLayers = editorState.selectedLayers;
  
  if (selectedLayers.isEmpty) {
    return;
  }

  final Map<String, String> groupIdMap = {};
  final List<Layer> copiedLayers = [];
  
  for (int i = 0; i < selectedLayers.length; i++) {
    final layer = selectedLayers[i];
    
    final newPosition = Offset(
      layer.offset.dx + 50,
      layer.offset.dy + 50,
    );
    
    final copiedLayer = _copyLayerWithPosition(layer, newPosition);
    if (copiedLayer != null) {
      
      if (layer.groupId != null) {
        String newGroupId;
        if (groupIdMap.containsKey(layer.groupId)) {
          newGroupId = groupIdMap[layer.groupId]!;
        } else {
          newGroupId = generateUniqueId();
          groupIdMap[layer.groupId!] = newGroupId;
        }
        copiedLayer.groupId = newGroupId;
      }
      
      copiedLayers.add(copiedLayer);
    }
  }

  for (int i = 0; i < copiedLayers.length; i++) {
    final layer = copiedLayers[i];
    editorState.addLayer(layer, autoCorrectZoomScale: false, autoCorrectZoomOffset: false);
  }

  editorState.layerInteractionManager.selectedLayerIds.clear();
  for (final layer in copiedLayers) {
    editorState.layerInteractionManager.selectedLayerIds.add(layer.id);
  }
}
```

[Screencast_20250725_174537.webm](https://github.com/user-attachments/assets/d02fa465-0323-411b-9cc9-c753e4a9a452)

hm21 committed:
fix(layer-selection): streamline layer tap handling and group selection logic

hm21 commented:
Ah, I see. I guess the problem you're having comes from the fractionalOffset applied to the text layer. Do you have this problem with other layers, such as between PaintLayers, WidgetLayers, and EmojiLayers?

saif-ellafi commented:
True, seems to be with Text Layers, I am investigating! but definitely, it is a matter of difference already in individual copy-paste.

Perhaps a bit off-topic, sorry! but it might mean we don't have coordinate problems with multi-select :)

hm21 committed:
fix(layer-selection): prevent group selection for layers without a groupId

saif-ellafi commented:
Just tested! Rock n' Roll, amazing teamwork!! Thanks for keeping me rolling, I hope it helped too!
Also copy-pasting good after  adding an offset to text layers. So, nice!

saif-ellafi commented:
Just _perhaps_ the interaction.remove() could delete all selected layers, but not suuuuure, to be honest, for a start it might be enough as is.

saif-ellafi commented:
Good thing about the whole refactor is that if also ever a draw-rectangle-to-select is implemented, this PR should help in now supporting multiple selected layers, so even if not perfect, it goes in the direction.

hm21 committed:
fix(layer-grouping-example): change interaction button action from edit to remove

hm21 commented:
Yeah was a quick speedrun to implement that feature but i also think that feature could be really usefull for a lot users and also for my projects.

I also thinked about `interaction.remove()` should remove all or not but then i thinked maybe confusing when like rotating/scaling work for just the layer itself and remove for all. What all layers will delete is when you move all selected layers into the default remove area. However i also think we let it for the moment and depending on feedback we can also add a config flag where users can enable it if they want.

> Good thing about the whole refactor is that if also ever a draw-rectangle-to-select is implemented...

To be honest, by reflex I already tried two or three times to draw a rectangle to select multiple layers, until I realized we haven’t implemented that yet😅

hm21 commented:
A prerelease of that feature is now available: `pro_image_editor: 11.0.0-dev.1`

saif-ellafi commented:
I'm pretty sure it is the first thing we'll hear from feedback

hm21 commented:
I will merge that PR into the new beta branch now. After a few days, if there are no issue reports from the prerelease, I will merge it into the stable branch :)

saif-ellafi commented:
Amazing, I'll be testing all through.
Everything is fine to be broken except layer positioning and history restore for me xD

I still live with big fear for the fractional offset but I feel we'll be fine since worst case some math can always be applied to fix I guess.

Thanks for everything!

hm21 commented:
To be honest, I'm pretty confident that `fractionalOffset` shouldn't cause major issues anymore. But as you said, in the worst case, a bit of math could fix it :)

FYI: I just released version `11.0.0-dev.3`, which includes two major changes:

1. I made significant performance improvements, which slightly change how filters appear. In my case, the difference was only noticeable when comparing them side by side, but depending on how many are applied, the effect could be more noticeable. Still, this change is important because many users messaged me recently saying the app crashes on mobile devices when too many changes are applied. So, in short, I merged all `Matrix4` transformations and color filters, which greatly reduces GPU usage.

2. I couldn't resist implementing layer drag selection #584 😉 There were a few challenges with rotated layers, but I think it's working well now. 
One important note: if the editor is zoomable, this feature is disabled because users can't move the editor content when zoomed in. However, with the new `enableMultiSelectMode` flag inside the MainEditor, it disables moving the editor and enables drag-to-select. So if you want to use it, just add a button that users can tap to toggle it on or off.
A simple preview of the feature is below.



<img width="400" alt="image" src="https://github.com/user-attachments/assets/f1143a9c-3450-4261-bec7-ec161a63ba33" />

saif-ellafi commented:
Woah @hm21 that is fantastic!

Some thoughts regarding disabling it on zoom and moving or selecting, because this is too good to have it on a button toggle!

What about:
1. Start moving with middle mouse button and/or right-click (also would solve https://github.com/hm21/pro_image_editor/issues/578) because really, moving with the same click that selects layers is problematic in practice. I checked for example: https://excalidraw.com/ or https://www.tldraw.com/ you'll notice most whiteboards move with middle-mouse button, and left mouse click is left exclusively for selection.
2. For Mobile, excalidraw moves with "double-finger drag", but taps are exclusive to select and multi-select (yes even to draw a rectangle!). Nothing to do here because you're already doing that, other than disable tap to move.

Moving is really important in an editor, so I suggest for the maturity of PIE to consider this (at best a setting to enable move with tap/left-click but disabled by default)

alternatively, although more ugly, would be to enable rectangular selection with shift+leftClick but this really is unnatural, just a lazy solution if you ask me :)

For mobile, you already zoom in and out with double pinch, so nothing to do there! Just using the tap to select and nothing more.

TLDR: If you disable left mouse / simple tap to move, many problems will be solved, and you can add rectangular selection by default instead!

What do you think?

saif-ellafi commented:
@hm21 Loving it! some early testing feedback shows that the exaggerated Paddings your worked on in the last release makes the rectangular area select more than it should.

I am experimenting with changing the mouse click functions in the meantime.

hm21 commented:
@saif-ellafi Yes, I also think using another mouse button is the way to go. For laptops, I think it will be useful to use the right mouse click to move. However, I also want to add some configurations that allow the user to choose which key performs which action.

I have already implemented a minimal version of this. [commit](https://github.com/hm21/pro_image_editor/commit/cb0f7411e8e6ea3a7189997defee991998319444).
Keep in mind it's not finished. I still need to add the configurations, and there is a problem where right-clicking on a layer allows you to move that layer. I also found that the drag selection is incorrect when zoomed in to the editor.

hm21 commented:
Btw some apps, like Figma, also allow you to hold the space bar to move around with the left mouse. So, that might be another option, in addition to the middle and right mouse buttons.

saif-ellafi commented:
I'll send a PR I am playing around, and I think it is really good already!

Regarding the rectangular selection, i don't think the padding is the problem, rather the current zoom in level. When I zoom in, it is much more inaccurate.
